README.md
ref-wchar
==========
### A ref "type" implementation of `wchar_t *` (a.k.a. wide string) backed by "node-iconv"
[![Build Status](https://secure.travis-ci.org/TooTallNate/ref-wchar.svg)](https://travis-ci.org/TooTallNate/ref-wchar)
[![Build Status](https://ci.appveyor.com/api/projects/status/xmne006y9jma3pm0?svg=true)](https://ci.appveyor.com/project/TooTallNate/ref-wchar)

This module offers a ["wide
strings"](http://en.wikipedia.org/wiki/Wide_character#C.2FC.2B.2B) (`wchar_t *`)
implementation on top of Node.js Buffers using the ref "type" interface.


Installation
------------

Install with `npm`:

``` bash
$ npm install ref-wchar
```


Examples
--------

Say you have a C library that exports a `wchar_t` char and a `wchar_t *` wide
string:

``` c
#if defined(WIN32) || defined(_WIN32)
#define EXPORT __declspec(dllexport)
#else
#define EXPORT
#endif

EXPORT whcar_t w = L'W';

EXPORT wchar_t s[] = L"hello world";

EXPORT wchar_t **str = (wchar_t **)(&s);
```

``` js
var ref = require('ref');
var dlfcn = require('dlfcn');
var wchar_t = require('ref-wchar');
var wchar_string = wchar_t.string;

var lib = dlfcn('./libexample');

// get the "w" symbol as a wchar
var w = lib.get('w', wchar_t.size);
ref.get(w, wchar_t);
// "W"

// get the "s" symbol as a wide string
var s = ref.reinterpretUntilZeros(lib.get('s'), wchar_t.size);
wchar_t.toString(s);
// "hello world"

// get the "str" pointer symbol as a wide string
var str = lib.get('str', wchar_string.size);
ref.get(str, wchar_string);
// "hello world"
```


License
-------

(The MIT License)

Copyright (c) 2014 Nathan Rajlich &lt;nathan@tootallnate.net&gt;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.md

1.0.2 / 2016-06-07
==================

  * Update travis.yml to use node version 6 (#6, @cookiengineer)
  * Update appveyor.xml to use node 6 as build target (#5, @cookiengineer)
  * package: update `iconv` dependency to use only major version 2 (#4, @cookiengineer)

1.0.1 / 2015-12-27
==================

  * package: update "ref" to v1.3.1
  * update for nan v2 / node.js v4
  * test: moar tests
  * test: test node v5
  * README: use SVG for appveyor badge

1.0.0 / 2015-03-24
==================

  * add .travis.yml and appveyor.yml files
  * index: use `toString()` in `wchar_t.get()`
  * README: add travis and appveyor badges
  * README: use C-style cast since we say C library
  * README: better example
  * package: update "ref" to v1.0.1

0.0.1 / 2014-06-19
==================

  * add initial tests
  * add .gitignore file
  * index: add lower-level `toString()` function
  * remove test.js file
  * index: throw a TypeError when a bad value is passed in to wchar_t set()
  * index: make `wchar_t` type based off of int16 or int32
  * index: add base `wchar_t` type
  * add README.md
  * beginnings of turning into a proper npm module


test.c
var ffi = require('ffi')
var assert = require('assert')
var wchar_t = require('./wchar_t')

// now we can create our FFI'd "wcslen()" function
var wcslenPtr = ffi.DynamicLibrary('libc.dylib').get('wcslen')
var wcslen = ffi.ForeignFunction(wcslenPtr, 'size_t', [ wchar_t ])

// "wcstod()" function
var wcstodPtr = ffi.DynamicLibrary('libc.dylib').get('wcstod')
var wcstod = ffi.ForeignFunction(wcstodPtr, 'double', [ wchar_t, ref.refType(wchar_t) ])

// you pass regular JavaScript Strings to the FFI'd function and the "wchar_t"
// type we defined handles converting it to a proper Buffer behind the scenes.
assert.equal(5, wcslen('hello'))
assert.equal(11, wcslen('hello world'))

assert.equal(-3.14, wcstod('-3.14', null));

test.js
var ffi = require('ffi');
var ref = require('ref');
var assert = require('assert');
var wchar_t = require('./');

// now we can create our FFI'd "wcslen()" function
var libc = ffi.Library('libc', {
  wcslen: [ 'size_t', [ wchar_t ] ],
  wcstod: [ 'double', [ wchar_t, ref.refType(wchar_t) ] ]
});

// you pass regular JavaScript Strings to the FFI'd function and the "wchar_t"
// type we defined handles converting it to a proper Buffer behind the scenes.
assert.equal(5, libc.wcslen('hello'));
assert.equal(11, libc.wcslen('hello world'));
assert.equal(-3.14, libc.wcstod('-3.14', null));


test/test.cc
#include <stdlib.h>
#include <wchar.h>
#include "nan.h"

using namespace node;

namespace {

static wchar_t w = L'w';
static wchar_t s[] = L"hello world";
static wchar_t **str = reinterpret_cast<wchar_t **>(&s);

void Initialize(v8::Handle<v8::Object> target) {
  Nan::HandleScope scope;

  Nan::Set(target, Nan::New<v8::String>("w").ToLocalChecked(), Nan::NewBuffer(reinterpret_cast<char *>(&w), sizeof(w)).ToLocalChecked());
  Nan::Set(target, Nan::New<v8::String>("s").ToLocalChecked(), Nan::NewBuffer(reinterpret_cast<char *>(&s), sizeof(s)).ToLocalChecked());
  Nan::Set(target, Nan::New<v8::String>("str").ToLocalChecked(), Nan::NewBuffer(reinterpret_cast<char *>(&str), sizeof(str)).ToLocalChecked());
}

} // anonymous namespace

NODE_MODULE(test, Initialize);


wchar_t.js
/**
 * Module dependencies.
 */

var ref = require('ref')
var Iconv = require('iconv').Iconv

// vars used by the "wchar_t" type
var wchar_size, getter, setter

// On Windows they're UTF-16 (2-bytes), but on Unix platform they're UTF-32
// (4-bytes).
if ('win32' == process.platform) {
  wchar_size = 2
  getter = new Iconv('UTF-16' + ref.endianness, 'UTF-8')
  setter = new Iconv('UTF-8', 'UTF-16' + ref.endianness)
} else {
  wchar_size = 4
  getter = new Iconv('UTF-32' + ref.endianness, 'UTF-8')
  setter = new Iconv('UTF-8', 'UTF-32' + ref.endianness)
}

// Create a "wchar_t *" type. We use the "CString" type as a base since it's pretty
// close to what we actually want. We just have to define custom "get" and "set"
// functions and then we can use this type in FFI functions.
var wchar_t = Object.create(ref.types.CString)
wchar_t.get = function get (buf, offset) {
  var _buf = buf.readPointer(offset)
  if (_buf.isNull()) {
    return null
  }
  var stringBuf = _buf.reinterpretUntilZeros(wchar_size)
  return getter.convert(stringBuf).toString('utf8')
};
wchar_t.set = function set (buf, offset, val) {
  var _buf = val // val is a Buffer? it better be \0 terminated...
  if ('string' == typeof val) {
    _buf = setter.convert(val + '\0')
  }
  return buf.writePointer(_buf, offset)
};

index.js

/**
 * Module dependencies.
 */

var ref = require('ref');
var Iconv = require('iconv').Iconv;

/**
 * On Windows they're UTF-16 (2-bytes),
 * but on Unix platform they're UTF-32 (4-bytes).
 *
 * TODO: add a way to optionally enable `-fshort-wchar` for Unix (gcc option).
 */

var size;
if ('win32' == process.platform) {
  size = 2;
} else {
  size = 4;
}

var getter = new Iconv('UTF-' + (8 * size) + ref.endianness, 'UTF-8');
var setter = new Iconv('UTF-8', 'UTF-' + (8 * size) + ref.endianness);


/**
 * The `wchar_t` type.
 */

exports = module.exports = Object.create(ref.types['int' + (8 * size)]);
exports.name = 'wchar_t';
exports.size = size;
exports.indirection = 1;
exports.get = function get (buf, offset) {
  if (offset > 0 || buf.length !== exports.size) {
    offset = offset | 0;
    buf = buf.slice(offset, offset + size);
  }
  return exports.toString(buf);
};
exports.set = function set (buf, offset, val) {
  var _buf = val; // assume val is a Buffer by default
  if (typeof val === 'string') {
    _buf = setter.convert(val[0]);
  } else if (typeof val === 'number') {
    _buf = setter.convert(String.fromCharCode(val));
  } else if (!_buf) {
    throw new TypeError('muss pass a String, Number, or Buffer for `wchar_t`');
  }
  return _buf.copy(buf, offset, 0, size);
};


/**
 * The "wchar_t *" type.
 *
 * We use the "CString" type as a base since it's pretty close to what we
 * actually want. We just have to define custom "get" and "set" functions.
 */

exports.string = Object.create(ref.types.CString);
exports.string.name = 'WCString';
exports.string.get = function get (buf, offset) {
  var _buf = buf.readPointer(offset);
  if (_buf.isNull()) {
    return null;
  }
  var stringBuf = _buf.reinterpretUntilZeros(exports.size);
  return exports.toString(stringBuf);
};
exports.string.set = function set (buf, offset, val) {
  var _buf = val; // val is a Buffer? it better be \0 terminated...
  if ('string' == typeof val) {
    _buf = setter.convert(val + '\0');
  }
  return buf.writePointer(_buf, offset);
};

/**
 * Turns a `wchar_t *` Buffer instance into a JavaScript String instance.
 *
 * @param {Buffer} buffer - buffer instance to serialize
 * @public
 */

exports.toString = function toString (buffer) {
  return getter.convert(buffer).toString('utf8');
};
